x/contractmanager/keeper/sudo.go

```diff
@@ -20,118 +20,37 @@ func (k Keeper) HasContractInfo(ctx sdk.Context, contractAddress sdk.AccAddress)
-			"error", err, "request", request, "contract_address", senderAddress)
-		return nil, fmt.Errorf("failed to marshal MessageResponse: %v", err)
-	}
-
-	resp, err := k.wasmKeeper.Sudo(ctx, senderAddress, m)
-	if err != nil {
-		k.Logger(ctx).Debug("SudoResponse: failed to Sudo",
-			"error", err, "contract_address", senderAddress)
-		return nil, fmt.Errorf("failed to Sudo: %v", err)
-	}
-
-	return resp, nil
-}
-
-func (k Keeper) SudoTimeout(
-	ctx sdk.Context,
-	senderAddress sdk.AccAddress,
-	request channeltypes.Packet,
-) ([]byte, error) {
-	k.Logger(ctx).Info("SudoTimeout", "senderAddress", senderAddress, "request", request)
-
-	if !k.wasmKeeper.HasContractInfo(ctx, senderAddress) {
-		if request.SourcePort == types.TransferPort {
-			// we want to allow non contract account to send the assets via IBC Transfer module
-			// we can determine the originating module by the source port of the request packet
-			return nil, nil
+	} else if ack != nil {
+		m.Error = &types.ErrorSudoPayload{
+			Request: request,
+			Details: ack.GetError(),
 		}
-		k.Logger(ctx).Debug("SudoTimeout: contract not found", "senderAddress", senderAddress)
-		return nil, fmt.Errorf("%s is not a contract address and not the Transfer module", senderAddress)
+	} else {
+		m.Timeout = &types.TimeoutPayload{Request: request}
 	}
-
-	x := types.MessageTimeout{}
-	x.Timeout.Request = request
-	m, err := json.Marshal(x)
+	data, err := json.Marshal(m)
 	if err != nil {
-		k.Logger(ctx).Error("failed to marshal MessageTimeout message",
-			"error", err, "request", request, "contract_address", senderAddress)
-		return nil, fmt.Errorf("failed to marshal MessageTimeout: %v", err)
+		return nil, fmt.Errorf("failed to marshal MessageSudoCallback: %v", err)
 	}
-
-	k.Logger(ctx).Info("SudoTimeout sending request", "data", string(m))
-
-	resp, err := k.wasmKeeper.Sudo(ctx, senderAddress, m)
-	if err != nil {
-		k.Logger(ctx).Debug("SudoTimeout: failed to Sudo",
-			"error", err, "contract_address", senderAddress)
-		return nil, fmt.Errorf("failed to Sudo: %v", err)
-	}
-
-	return resp, nil
+	return data, nil
 }
 
-func (k Keeper) SudoError(
-	ctx sdk.Context,
-	senderAddress sdk.AccAddress,
-	request channeltypes.Packet,
-	details string,
-) ([]byte, error) {
-	k.Logger(ctx).Debug("SudoError", "senderAddress", senderAddress, "request", request)
-
-	if !k.wasmKeeper.HasContractInfo(ctx, senderAddress) {
-		if request.SourcePort == types.TransferPort {
-			// we want to allow non contract account to send the assets via IBC Transfer module
-			// we can determine the originating module by the source port of the request packet
-			return nil, nil
-		}
-		k.Logger(ctx).Debug("SudoError: contract not found", "senderAddress", senderAddress)
-		return nil, fmt.Errorf("%s is not a contract address and not the Transfer module", senderAddress)
+func PrepareOpenAckCallbackMessage(details types.OpenAckDetails) ([]byte, error) {
```

Comment:
the `PrepareSudoCallbackMessage` func also seems not at the right place for me, but I couldn't come up with a better one since it's used by two modules (not by a single one as `PrepareOpenAckCallbackMessage`). but since they are quite similar, I think we can either keep the both here or find better places for both of them. a better place for `PrepareOpenAckCallbackMessage` is obvious — `interchaintxs` module. any ideas on `PrepareSudoCallbackMessage`?